docs/source/peft_integration.md

```diff
@@ -1,144 +1,671 @@
+- `--use_rslora`: Enable Rank-Stabilized LoRA
+- `--use_dora`: Enable Weight-Decomposed LoRA (DoRA)
+- `--load_in_4bit`: Enable 4-bit quantization (QLoRA)
+- `--load_in_8bit`: Enable 8-bit quantization
+
+## Saving and Loading PEFT Models
+
+### Saving
+
+After training, save your PEFT adapters:
+
+```python
+# Save the adapters
+trainer.save_model("path/to/adapters")
+
+# Or manually
+model.save_pretrained("path/to/adapters")
+```
+
+This saves only the adapter weights (~few MB) rather than the full model (~several GB).
+
+### Loading
+
+Load a PEFT model for inference:
+
+```python
+from transformers import AutoModelForCausalLM
+from peft import PeftModel
+
+# Load base model
+base_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B")
+
+# Load PEFT adapters
+model = PeftModel.from_pretrained(base_model, "path/to/adapters")
+
+# Optionally merge adapters into base model for faster inference
+model = model.merge_and_unload()
 ```
 
-## Fine-tuning Llama-2 model
+### Pushing to Hub
 
-You can easily fine-tune Llama2 model using `SFTTrainer` and the official script! For example to fine-tune llama2-7b on the Guanaco dataset, run (tested on a single NVIDIA T4-16GB):
+You can easily share your PEFT adapters on the Hugging Face Hub:
+
+```python
+# Push adapters to Hub
+model.push_to_hub("username/model-name-lora")
+
+# Load from Hub
+from peft import PeftModel
+model = PeftModel.from_pretrained(base_model, "username/model-name-lora")
+```
+
+## Multi-GPU Training
+
+PEFT works seamlessly with TRL's multi-GPU support through `accelerate`:
 
 ```bash
-python trl/scripts/sft.py --output_dir sft_openassistant-guanaco  --model_name meta-llama/Llama-2-7b-hf --dataset_name timdettmers/openassistant-guanaco --load_in_4bit --use_peft --per_device_train_batch_size 4 --gradient_accumulation_steps 2
+# Configure accelerate
+accelerate config
+
+# Launch training
+accelerate launch trl/scripts/sft.py \
+    --model_name_or_path Qwen/Qwen2-0.5B \
+    --dataset_name trl-lib/Capybara \
+    --use_peft \
+    --lora_r 32 \
+    --lora_alpha 16
 ```
+
+For QLoRA with multiple GPUs, the base model is automatically sharded:
+
+```bash
+accelerate launch trl/scripts/sft.py \
+    --model_name_or_path meta-llama/Llama-2-70b-hf \
+    --load_in_4bit \
+    --use_peft \
+    --lora_r 32
+```
+
+## Troubleshooting
+
+### Out of Memory Errors
+
+If you encounter OOM errors:
+
+1. Enable QLoRA: `--load_in_4bit`
+2. Reduce batch size: `--per_device_train_batch_size 1`
+3. Increase gradient accumulation: `--gradient_accumulation_steps 16`
+4. Enable gradient checkpointing: `--gradient_checkpointing`
+5. Reduce LoRA rank: `--lora_r 8`
+6. Reduce target modules: `--lora_target_modules q_proj v_proj`
+
+### Slow Training
+
+If training is slow:
+
+1. Increase batch size (if memory allows)
+2. Use Flash Attention 2: `--attn_implementation flash_attention_2`
+3. Use bf16: `--bf16`
+4. Reduce gradient checkpointing frequency
+
+
+
```

Comment:
Most a these are not specific to peft, so I recommend removing this section, and add these elements in `reducing_memory_usage.md` or `speeding_up_training.md` (can be done in a follow-up PR)